### PR TITLE
FT cleanup

### PR DESCRIPTION
**Pull request by shufay**
Cleaned up legacy thermal AFQMC code to mirror the structure of 0T code.

### CODE REVIEW

File: ipie/addons/thermal/walkers/tests/test_thermal_walkers.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+        numpy.testing.assert_almost_equal(legacy_eloc, eloc, decimal=10)
+        numpy.testing.assert_almost_equal(legacy_walkers.walkers[iw].G[0], walkers.Ga[iw], decimal=10)
+        numpy.testing.assert_almost_equal(legacy_walkers.walkers[iw].G[1], walkers.Gb[iw], decimal=10)
+        numpy.testing.assert_almost_equal(legacy_walkers.walkers[iw].stack.ovlp[0], walkers.stack[iw].ovlp[0], decimal=10)
+        numpy.testing.assert_almost_equal(legacy_walkers.walkers[iw].stack.ovlp[1], walkers.stack[iw].ovlp[1], decimal=10)
+
+
+#@pytest.mark.skipif(_no_cython, reason="Need to build cython modules.")
+#@pytest.mark.unit
```
Comment (linusjoonho) [inline ipie/addons/thermal/walkers/tests/test_thermal_walkers.py:116]: why is it not marked?

File: ipie/qmc/options.py [hunk 2]
```diff
@@ -148,6 +157,18 @@ def __init__(self, inputs={}, verbose=False):
             alias=["random_seed", "seed"],
             verbose=verbose,
         )
+        self.mu = get_input_value(
```
Comment (linusjoonho) [inline ipie/qmc/options.py:160]: Create ThermalQMCOpts and move it there?

File: ipie/qmc/options.py [hunk 3]
```diff
@@ -148,6 +157,18 @@ def __init__(self, inputs={}, verbose=False):
...
+            inputs,
+            "mu",
+            default=None,
+            verbose=verbose,
+        )
+        self.beta = get_input_value(
```
Comment (linusjoonho) [inline ipie/qmc/options.py:166]: Create ThermalQMCOpts and move it there?

File: ipie/trial_wavefunction/single_det.py [hunk 4]
```diff
@@ -40,7 +40,7 @@ def __init__(self, wavefunction, num_elec, num_basis, handler=MPIHandler(), verb
         self._max_num_dets = 1
         imag_norm = numpy.sum(self.psi.imag.ravel() * self.psi.imag.ravel())
         if imag_norm <= 1e-8:
-            # print("# making trial wavefunction MO coefficient real")
+            #print("# making trial wavefunction MO coefficient real")
```
Comment (linusjoonho) [inline ipie/trial_wavefunction/single_det.py:43]: delete?

File: ipie/trial_wavefunction/single_det.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def __init__(self, wavefunction, num_elec, num_basis, handler=MPIHandler(), verb
         self._max_num_dets = 1
         imag_norm = numpy.sum(self.psi.imag.ravel() * self.psi.imag.ravel())
         if imag_norm <= 1e-8:
```
Comment (linusjoonho) [inline ipie/trial_wavefunction/single_det.py:42]: delete this if statement?

File: ipie/trial_wavefunction/tests/test_single_det_trial.py [hunk 6]
```diff
@@ -40,7 +40,7 @@ def test_single_det():
     assert trial._rH1b.shape == (nbeta, nbasis)
 
 
-@pytest.mark.unit
+#@pytest.mark.unit
```
Comment (linusjoonho) [inline ipie/trial_wavefunction/tests/test_single_det_trial.py:43]: secretly removing a test, why?

File: ipie/walkers/pop_controller.py [hunk 7]
```diff
@@ -138,7 +138,7 @@ def get_buffer(walkers, iw):
     buff = xp.zeros(walkers.buff_size, dtype=numpy.complex128)
     for d in walkers.buff_names:
         data = walkers.__dict__[d]
-        if data is None:
+        if (data is None) or isinstance(data, (int, float, complex, numpy.float64, numpy.complex128)):
```
Comment (linusjoonho) [inline ipie/walkers/pop_controller.py:141]: Create ThermalPopController by inheriting from PopController. Localize the changes into addons so that the main pop_controller.py does not depend on addons.

**Review by linusjoonho — COMMENTED** [review]
Please try to address the comments. It looks good overall.

File: ipie/addons/thermal/walkers/tests/test_thermal_walkers.py (hunk 1 above)
Comment (shufay) [reply ipie/addons/thermal/walkers/tests/test_thermal_walkers.py:116]: lowrank code is not working--should I remove?

File: ipie/addons/thermal/propagation/phaseless_base.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+from ipie.addons.thermal.propagation.force_bias import construct_force_bias
+from ipie.addons.thermal.propagation.operations import apply_exponential
+from ipie.utils.backend import arraylib as xp
+from ipie.hamiltonians.generic import GenericRealChol, GenericComplexChol
+
+# TODO: Add lowrank implementation.
+# Ref: 10.1103/PhysRevB.80.214116 for bounds.
+
+@plum.dispatch
+def construct_mean_field_shift(hamiltonian: GenericRealChol, trial):
```
[the remark this answers is not in the document]
Comment (shufay) [reply ipie/addons/thermal/propagation/phaseless_base.py:20]: zero temperature functions use `trial.G`. If we stored `P` as `trial.G` I think it would be confusing.

File: ipie/addons/thermal/propagation/operations.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+from ipie.utils.backend import arraylib as xp
+
+def apply_exponential(VHS, exp_nmax):
```
[the remark this answers is not in the document]
Comment (shufay) [reply ipie/addons/thermal/propagation/operations.py:3]: removed this function

File: ipie/addons/thermal/estimators/thermal.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        # C = (Db Q^{-1}T^{-1}+Ds)
+        C = numpy.dot(numpy.einsum("ii,ij->ij", Db, Q1.conj().T), T1inv) + Ds
+        Cinv = scipy.linalg.inv(C, check_finite=False)
+
+        # Then G = T^{-1} C^{-1} Db Q^{-1}
+        # Q is unitary.
+        G.append(numpy.dot(numpy.dot(T1inv, Cinv), numpy.einsum("ii,ij->ij", Db, Q1.conj().T)))
+    return one_rdm_from_G(numpy.array(G))
+
+def entropy(beta, mu, H):
```
[the remark this answers is not in the document]
Comment (shufay) [reply ipie/addons/thermal/estimators/thermal.py:74]: no****

File: ipie/addons/thermal/estimators/thermal.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+    Returns
+    -------
+    P : :class:`numpy.ndarray`
+        Thermal 1RDM.
+    """
+    I = numpy.identity(G.shape[-1])
+    return numpy.array([I - G[0].T, I - G[1].T], dtype=numpy.complex128)
+
+def one_rdm_stable(BT, num_slices):
```
[the remark this answers is not in the document]
Comment (shufay) [reply ipie/addons/thermal/estimators/thermal.py:26]: yes--used in `trial/*`

File: ipie/addons/thermal/estimators/energy.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        walkers: UHFThermalWalkers):
+    energies = xp.zeros((walkers.nwalkers, 3), dtype=xp.complex128)
+    
+    for iw in range(walkers.nwalkers):
+        # Want the full Green's function when calculating observables.
+        walkers.calc_greens_function(iw, slice_ix=walkers.stack[iw].nslice)
+        P = one_rdm_from_G(xp.array([walkers.Ga[iw], walkers.Gb[iw]]))
+        energy = local_energy_generic_cholesky(hamiltonian, P)
+
+        for i in range(3):
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/energy.py:40]: just broadcast: energies[iw,:] = energy

File: ipie/addons/thermal/estimators/generic.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
+import plum
+import numpy
+from ipie.hamiltonians.generic import GenericRealChol, GenericComplexChol
+
+
+@plum.dispatch
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/generic.py:6]: Aren't these identical to the non thermal functions? P should have a type hint.

File: ipie/addons/thermal/estimators/handler.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
+"""Routines and classes for estimation of observables."""
+
+from __future__ import print_function
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/handler.py:3]: This is not used right?

File: ipie/addons/thermal/estimators/handler.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+            shift = None
+        walker_factors.eshift = comm.bcast(shift)
+        if comm.rank == 0:
+            self.output.push_to_chunk(self.global_estimates, f"data")
+            self.output.increment()
+        if comm.rank == 0:
+            print(f"{block:>17d} " + output_string)
+        self.zero()
+    
+    def print_cut(self, comm, block, t, walker_factors):
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/handler.py:221]: What does this function do?

File: ipie/addons/thermal/estimators/tests/test_estimators.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
+import pytest
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/tests/test_estimators.py:1]: Every file should have the license in it at the top.

File: ipie/addons/thermal/propagation/__init__.py
Comment (fdmalone) [inline ipie/addons/thermal/propagation/__init__.py:1]: Every file needs the license.

File: ipie/addons/thermal/utils/testing.py [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+from ipie.addons.thermal.qmc.options import ThermalQMCParams
+from ipie.addons.thermal.qmc.thermal_afqmc import ThermalAFQMC
+
+
+def build_generic_test_case_handlers(options: dict,
+                                     seed: Union[int, None],
+                                     debug: bool = False,
+                                     with_eri: bool = False,
+                                     verbose: int = 0):
+    # Unpack options
```
Comment (fdmalone) [inline ipie/addons/thermal/utils/testing.py:23]: I think it might be better to a) explicitly spell out the options in the dict as kwargs, you're not saving anything by unpacking them later, and it serves as form of documentation to have them as args/kwargs OR b) define little dataclasses which group the options so that you don't need to pass in 20 arguments.

File: ipie/trial_wavefunction/single_det.py (hunk 4 above)
Comment (shufay) [reply ipie/trial_wavefunction/single_det.py:43]: This change will be in a new branch

File: ipie/trial_wavefunction/single_det.py (hunk 5 above)
Comment (shufay) [reply ipie/trial_wavefunction/single_det.py:42]: This change will be in a new branch

File: ipie/addons/thermal/estimators/handler.py (hunk 15 above)
Comment (shufay) [reply ipie/addons/thermal/estimators/handler.py:221]: Prints the energy and particle number estimators at each time slice. It's used to generate the imaginary time trace of estimators

File: ipie/addons/thermal/estimators/generic.py (hunk 13 above)
Comment (shufay) [reply ipie/addons/thermal/estimators/generic.py:6]: I think the non-thermal `local_energy_generic_cholesky` (https://github.com/JoonhoLee-Group/ipie/blob/693f3dbf861f4ee75f9f923a5f7c26959864e85f/ipie/estimators/generic.py#L55) is wrong for the `GenericComplexChol` case. The value returned disagrees with the energy computed from ERIs. But I also don't think `local_energy_generic_cholesky` is used in the non-thermal code.

Comment (fdmalone) [reply ipie/addons/thermal/estimators/generic.py:24]: There's a bug for complex ERIs, or rather the linked function assumes real integrals. Could you open an issue?

File: ipie/addons/thermal/estimators/handler.py (hunk 15 above)
Comment (fdmalone) [reply ipie/addons/thermal/estimators/handler.py:237]: Cool. Maybe rename it print_time_slice, and add a docstring.

File: ipie/addons/thermal/estimators/energy.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        self._data = {
+            "ENumer": 0.0j,
+            "EDenom": 0.0j,
+            "ETotal": 0.0j,
+            "E1Body": 0.0j,
+            "E2Body": 0.0j,
+        }
+
+        self._shape = (len(self.names),)
+        self._data_index = {k: i for i, k in enumerate(list(self._data.keys()))}
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/energy.py:59]: What is this used for?

Comment (fdmalone) [reply ipie/addons/thermal/estimators/energy.py:59]: Nevermind I wrote the code lol.

File: ipie/addons/thermal/estimators/energy.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        # Want the full Green's function when calculating observables.
+        walkers.calc_greens_function(iw, slice_ix=walkers.stack[iw].nslice)
+        P = one_rdm_from_G(xp.array([walkers.Ga[iw], walkers.Gb[iw]]))
+        energy = local_energy_generic_cholesky(hamiltonian, P)
+        energies[iw] = energy
+
+    return energies
+
+
+class ThermalEnergyEstimator(EstimatorBase):
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/energy.py:45]: I think this is essentially identical to EnergyEstimator right? You could save some space and do:

``` python
class ThermalEnergyEstimator(EnergyEstimator):
    # Only overwrite the single method compute_estimator to have the right import.
```

File: ipie/addons/thermal/estimators/generic.py [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    X = Xa + Xb
+    ecoul = 0.5 * numpy.dot(X, X)
+    
+    # Ex.
+    PaT = Pa.T.copy()
+    PbT = Pb.T.copy()
+    T = numpy.zeros((nbasis, nbasis), dtype=numpy.complex128)
+    exx = 0.0j  # we will iterate over cholesky index to update Ex energy for alpha and beta
+
+    for x in range(nchol):  # write a cython function that calls blas for this.
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/generic.py:60]: I don't think cython, maybe numba.

File: ipie/addons/thermal/estimators/handler.py [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+from ipie.utils.io import format_fixed_width_strings
+
+# Some supported (non-custom) estimators
+_predefined_estimators = {
+    "energy": ThermalEnergyEstimator,
+    "nav": ThermalNumberEstimator,
+}
+
+
+class ThermalEstimatorHandler(object):
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/handler.py:41]: From a glance I think this is essentialy identical to EstimatorHandler, with the only specialization being the print_time_slice method. You can use inheritance and reduce the amount of code here.

File: ipie/addons/thermal/estimators/local_energy.py [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+#
+# Authors: Fionn Malone <fmalone@google.com>
+#          Joonho Lee
+#
+
+from ipie.addons.thermal.estimators.generic import local_energy_generic_cholesky
+from ipie.addons.thermal.estimators.thermal import one_rdm_from_G
+
+def local_energy_P(hamiltonian, trial, P):
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/local_energy.py:22]: Add type hints to the arguments. Maybe rename local_energy_from_density_matrix.

File: ipie/addons/thermal/estimators/particle_number.py [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+# Authors: Fionn Malone <fmalone@google.com>
+#          Joonho Lee
+#
+
+from ipie.utils.backend import arraylib as xp
+from ipie.estimators.estimator_base import EstimatorBase
+from ipie.addons.thermal.estimators.thermal import one_rdm_from_G
+
+
+def particle_number(dmat):
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/particle_number.py:24]: type hints.

File: ipie/addons/thermal/estimators/tests/test_estimators.py [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+complex_integrals = False
+debug = True
+verbose = True
+seed = 7
+numpy.random.seed(seed)
+
+@pytest.mark.unit
+def test_energy_estimator():
+    # Test.
+    print('\n----------------------------')
```
Comment (fdmalone) [inline ipie/addons/thermal/estimators/tests/test_estimators.py:57]: Remove the prints from the added tests.

File: ipie/addons/thermal/qmc/calc.py [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+from ipie.hamiltonians.utils import get_hamiltonian
+
+from ipie.addons.thermal.trial.utils import get_trial_density_matrix
+from ipie.addons.thermal.walkers.uhf_walkers import UHFThermalWalkers
+from ipie.addons.thermal.propagation.propagator import Propagator
+from ipie.addons.thermal.qmc.options import ThermalQMCOpts, ThermalQMCParams
+from ipie.addons.thermal.qmc.thermal_afqmc import ThermalAFQMC
+
+
+def init_communicator():
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/calc.py:36]: Remove these duplicate functions, just import from ipie.

File: ipie/addons/thermal/qmc/options.py [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+        Number of steps before QR stabilization of walkers is performed.
+    pop_control_freq : int
+        Frequency at which population control occurs.
+    rng_seed : int
+        The random number seed. If run in parallel the seeds on other cores / 
+        threads are determined from this.
+    """
+    # Due to structure of FT algorithm, `num_steps_per_block` is fixed at 1.
+    # Overide whatever input for backward compatibility.
+    num_steps_per_block: ClassVar[float] = 1
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/options.py:113]: Why is this a float?

File: ipie/addons/thermal/qmc/options.py [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+    pop_control_freq : int
+        Frequency at which population control occurs.
+    rng_seed : int
+        The random number seed. If run in parallel the seeds on other cores / 
+        threads are determined from this.
+    """
+    # Due to structure of FT algorithm, `num_steps_per_block` is fixed at 1.
+    # Overide whatever input for backward compatibility.
+    num_steps_per_block: ClassVar[float] = 1 
+    mu: float = _no_default
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/options.py:114]: Why is this here?

Comment (fdmalone) [reply ipie/addons/thermal/qmc/options.py:114]: I think a more natural sentinel value would be None in this case. Then in __post_init__ check the value is not None. I don't think this is a hack in this case.

Comment (fdmalone) [reply ipie/addons/thermal/qmc/options.py:114]: For example, `beta: Optional[float] = None`, is the preferred way of doing this.

File: ipie/addons/thermal/qmc/options.py (hunk 26 above)
Comment (fdmalone) [reply ipie/addons/thermal/qmc/options.py:113]: Why is this ClassVar and not just int = 1

File: ipie/addons/thermal/qmc/thermal_afqmc.py [hunk 28]
```diff
@@ -0,0 +1,472 @@
...
+                    start = time.time()
+                    self.pcontrol.pop_control(self.walkers, comm)
+                    synchronize()
+                    self.tpopc += time.time() - start
+                    self.tpopc_send = self.pcontrol.timer.send_time
+                    self.tpopc_recv = self.pcontrol.timer.recv_time
+                    self.tpopc_comm = self.pcontrol.timer.communication_time
+                    self.tpopc_non_comm = self.pcontrol.timer.non_communication_time
+                
+                # Print cut.
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/thermal_afqmc.py:274]: Remove commented out code. Do you intend to make it a configurable option to print out time slice information?

File: ipie/addons/thermal/qmc/thermal_afqmc.py [hunk 29]
```diff
@@ -0,0 +1,472 @@
...
+            else:
+                eshift += self.accumulators.eshift - eshift
+
+            self.walkers.reset(self.trial) # Reset stack, weights, phase.
+
+            synchronize()
+            self.tpath += time.time() - start_path
+
+
+    def finalise(self, verbose=False):
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/thermal_afqmc.py:310]: If you didn't modify the timers just delete this and allow inheritance to do its job.

File: ipie/addons/thermal/qmc/thermal_afqmc.py [hunk 30]
```diff
@@ -0,0 +1,472 @@
...
+                    )
+                )
+                print(
+                    "#       - Non-Commnication: {:.6f} s / call for {} call(s) in each of {} blocks".format(
+                        self.tpopc_non_comm / (npcon * nblocks), npcon, nblocks
+                    )
+                )
+
+
+    def get_env_info(self):
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/thermal_afqmc.py:403]: Is this special for FT? If not remove and use inheritance.

File: ipie/addons/thermal/qmc/thermal_afqmc.py [hunk 31]
```diff
@@ -0,0 +1,472 @@
...
+
+        # Calculate estimates for initial distribution of walkers.
+        self.estimators.compute_estimators(self.hamiltonian,
+                                           self.trial, self.walkers)
+        self.accumulators.update(self.walkers)
+        self.estimators.print_block(comm, 0, self.accumulators)
+        self.accumulators.zero()
+
+
+    def setup_timers(self):
```
Comment (fdmalone) [inline ipie/addons/thermal/qmc/thermal_afqmc.py:460]: Again, remove if not specialized for FT.

File: ipie/addons/thermal/trial/one_body.py [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+import scipy.linalg
+
+from ipie.addons.thermal.estimators.particle_number import particle_number
+from ipie.addons.thermal.estimators.thermal import one_rdm_stable
+from ipie.addons.thermal.estimators.greens_function import greens_function
+from ipie.addons.thermal.trial.chem_pot import compute_rho, find_chemical_potential
+from ipie.utils.misc import update_stack
+
+
+class OneBody(object):
```
Comment (fdmalone) [inline ipie/addons/thermal/trial/one_body.py:29]: The Class(object) syntax is deprecated in python (I know this is a hangover from legacy.)

File: ipie/addons/thermal/utils/ueg.py [hunk 33]
```diff
@@ -0,0 +1,583 @@
...
+#
+# Authors: Fionn Malone <fmalone@google.com>
+#          Joonho Lee
+#
+
+import numpy
+import scipy.sparse
+from ipie.utils.io import write_qmcpack_sparse
+
+# -----------------------------------------------------------------------------
```
Comment (fdmalone) [inline ipie/addons/thermal/utils/ueg.py:23]: just `class UEG:` is fine. The # ---- docstrings don't add much here. Why is this in utils?

File: ipie/addons/thermal/utils/ueg.py [hunk 34]
```diff
@@ -0,0 +1,583 @@
...
+        Number of field configurations per walker for back propagation.
+
+    basis : :class:`numpy.ndarray`
+        Basis vectors within a cutoff.
+
+    kfac : float
+        Scale factor (2pi/L).
+    """
+
+    def __init__(self, options, verbose=False):
```
Comment (fdmalone) [inline ipie/addons/thermal/utils/ueg.py:70]: Again not your problem but we shouldn't pass in big dicts of parameters. Feel free to not address this now.

File: ipie/addons/thermal/walkers/pop_controller.py [hunk 35]
```diff
@@ -0,0 +1,442 @@
...
+            self.reconfiguration_counter += 1
+            if self.reconfiguration_counter % self.reconfiguration_freq == 0:
+                stochastic_reconfiguration(walkers, comm, self.timer)
+                self.reconfiguration_counter = 0
+        else:
+            if comm.rank == 0:
+                print("Unknown population control method.")
+
+
+def get_buffer(walkers, iw):
```
Comment (fdmalone) [inline ipie/addons/thermal/walkers/pop_controller.py:94]: Why is any of this code necessary? The population control algorithms should be agnostic to the underlying qmc algorithm. If it's not the case then that's a problem and we should open an issue. But the get/set buffers which are the only potentially algorithm dependent quantities should be delegated to the walkers right?

File: ipie/addons/thermal/walkers/uhf_walkers.py [hunk 36]
```diff
@@ -0,0 +1,169 @@
...
+        self.M0b = numpy.array([
+                    scipy.linalg.det(self.Gb[iw], check_finite=False) for iw in range(self.nwalkers)])
+
+        for iw in range(self.nwalkers):
+            self.stack[iw].ovlp = numpy.array([1.0 / self.M0a[iw], 1.0 / self.M0b[iw]])
+
+        # # Temporary storage for stacks...
+        # We should kill these here and store them in stack (10/02/2023)
+        # I = numpy.identity(self.nbasis, dtype=numpy.complex128)
+        # One = numpy.ones(self.nbasis, dtype=numpy.complex128)
```
Comment (fdmalone) [inline ipie/addons/thermal/walkers/uhf_walkers.py:125]: Delete commented out code.

File: ipie/trial_wavefunction/single_det.py [hunk 37]
```diff
@@ -78,6 +78,7 @@ def calculate_energy(self, system, hamiltonian) -> numpy.ndarray:
         self.energy = self.e1b + self.e2b
 
         if self.verbose:
+            print("# (Ej, Ek) = {}, {}".format(self.ej, self.ek))
```
Comment (fdmalone) [inline ipie/trial_wavefunction/single_det.py:81]: ?

**Review by fdmalone — CHANGES_REQUESTED** [review]
Thanks for this! I know this must have been very painful.

File: ipie/addons/thermal/trial/tests/test_chem_pot.py [hunk 38]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2022 The ipie Developers. All Rights Reserved.
```
Comment (fdmalone) [inline ipie/addons/thermal/trial/tests/test_chem_pot.py:1]: You need an __init__.py in the the test folders too

File: ipie/addons/thermal/utils/ueg.py (hunk 33 above)
Comment (shufay) [reply ipie/addons/thermal/utils/ueg.py:23]: The `ueg.py` was only meant for testing. I think the idea was to add a UEG example for users to refer to. I can add that.

Comment (fdmalone) [reply ipie/addons/thermal/utils/ueg.py:23]: Right, if it's just an example I would put it in examples (not utils). If it's a bone-fide system it should go in Hamiltonian.

File: ipie/addons/thermal/qmc/thermal_afqmc.py (hunk 28 above)
Comment (shufay) [reply ipie/addons/thermal/qmc/thermal_afqmc.py:274]: Sure--if you think it's a good idea?

File: ipie/addons/thermal/qmc/options.py (hunk 26 above)
Comment (shufay) [reply ipie/addons/thermal/qmc/options.py:113]: I think I did this so you can't assign an instance of `ThermalQMCParams` the `num_steps_per_block` value, as opposed to `int = 1` which sets a default.

Comment (fdmalone) [reply ipie/addons/thermal/qmc/options.py:113]: Makes sense, but it should be ClassVar[int] (not float).

File: ipie/addons/thermal/walkers/pop_controller.py (hunk 35 above)
Comment (fdmalone) [reply ipie/addons/thermal/walkers/pop_controller.py:94]: Right, the walker should do the concatenation. Can you open an issue to track this.

**Comment by fdmalone** [comment]
The test is failing because you need to build the cython modules. I'm uneasy that this cythonizing is again becoming necessary. Can you include an import guard so that the tests run without the need for cythonizing (you should see in the test output that several tests are already skipped due to missing cythonized modules.

File: ipie/addons/thermal/qmc/thermal_afqmc.py (hunk 28 above)
Comment (fdmalone) [reply ipie/addons/thermal/qmc/thermal_afqmc.py:274]: I think it could be a useful diagnostic in the future. You don't need to do it now. Just open an issue to track.

File: ipie/addons/thermal/estimators/handler.py (hunk 21 above)
Comment (shufay) [reply ipie/addons/thermal/estimators/handler.py:41]: The `thermal/estimators/` folder has no dependence on the non-thermal `systems` classes, so I'll just keep the methods that don't call `systems` but do in the non-thermal code.

File: ipie/addons/thermal/utils/ueg.py (hunk 33 above)
Comment (shufay) [reply ipie/addons/thermal/utils/ueg.py:23]: It's called in `utils/testing.py` though--do we want it in the non-thermal `hamiltonian`? I thought it was discussed to only have `generic*.py` there at one point

File: ipie/addons/thermal/qmc/thermal_afqmc.py (hunk 28 above)
Comment (shufay) [reply ipie/addons/thermal/qmc/thermal_afqmc.py:274]: Added a keyword argument to `ThermalAFQMC.run`

**Comment by shufay** [comment]
Thanks for the feedback and looking over it all. I've tried to address all your comments--please let me know if there's anything else I should look at!

File: ipie/addons/thermal/utils/ueg.py (hunk 33 above)
Comment (fdmalone) [reply ipie/addons/thermal/utils/ueg.py]: I think if we just want it for testing the more sensible (larger refactor would be), make a folder called testing, move all the testing methods into their own separate modules in testing/* and make a testing/hamiltonians.py to add new test only hamiltonians. Utils should disappear largely speaking #242. For now I guess this is ok.

**Review by fdmalone — APPROVED** [review]

File: ipie/addons/thermal/estimators/energy.py [hunk 39]
```diff
@@ -42,9 +43,11 @@ def local_energy(
...
 
 
-class ThermalEnergyEstimator(EstimatorBase):
+class ThermalEnergyEstimator(EnergyEstimator):
     def __init__(self, hamiltonian=None, trial=None, filename=None):
-        super().__init__()
+        assert hamiltonian is not None
+        assert trial is not None
+        EstimatorBase.__init__(self)
         self._eshift = 0.0
```
[the remark this answers is not in the document]
Comment (shufay) [reply ipie/addons/thermal/estimators/energy.py:51]: The thermal `__init__` does not refer to `system` at all. If I try to pass in `system=None` into `ThermalEnergyEstimator` an error will be raised here:

https://github.com/JoonhoLee-Group/ipie/blob/693f3dbf861f4ee75f9f923a5f7c26959864e85f/ipie/estimators/energy.py#L121

Should this be changed then?

[the remark this answers is not in the document]
Comment (fdmalone) [reply ipie/addons/thermal/estimators/energy.py:51]: Yeah just remove the asserts in the EnergyEstimator class.

[the remark this answers is not in the document]
Comment (fdmalone) [reply ipie/addons/thermal/estimators/energy.py:51]: Keep the None kwargs for the moment.
